### ISSUE TITLE
API: Return history for a single check

### ISSUE THREAD

**Issue by jameskirsop** [healthchecks/healthchecks#349]
Would be great to allow a `?history=x` query variable, where `x` is the number of days you wish to retrieve historical data for to the single end point, so it's possible to show availability over time.

Will try and write this up in a new branch and do a PR.

**Comment by jameskirsop** [comment]
Realised that this is going to be less useful than I originally thought because there's no events stored for failures, unless there's alerting set up.

@cuu508, have you considered storing a failure event for checks that pass their grace time without a ping?

**Comment by cuu508** [comment]
To clarify, with the `?history=x` parameter, would you want the API return JSON with the same structure, but with values from x days ago? 

We're not tracking the edit history of things like check's name, tags, schedule etc., so this would be difficult to implement properly.

**Comment by jameskirsop** [comment]
No, I'm not wanting to track edit history, but status history.

To illustrate, at the moment, if you look at the status page for a single check (`/checks/<uuid>/details`) you are presented with a Log view. It's the data of this log view I would like to return via an additional 'history' attribute on the returned JSON object.

However, if you have no notifications setup, that Log view shows all the successful Pings, but none of the times where it's passed the grace period (i.e. failed according to the configured parameters for that check).

I'm suggesting that a new type of event be added to the _Log_ that show a non-ping failure event, not just when a notification is sent, but any time a failure occurs. Perhaps this _could_ be done by creating a new `Ping` object with the `fail` status when the grace period passes?

The `history` attribute on the returned JSON object (which would just be an addition to the existing attributes) would contain an ordered list of events with timestamps and event `kinds`, similar to the summary in the `details` view.

**Comment by jameskirsop** [comment]
I've [built something that fulfils](https://github.com/jameskirsop/healthchecks/tree/return-single-history) my needs by using a query on the `Flips` objects attached to a check.

These changes return a JSON object that looks like this:
```
{
    "name": "My First Check (copy)",
    "tags": "",
    "desc": "",
    "grace": 3600,
    "n_pings": 1,
    "status": "down",
    "last_ping": "2020-03-23T23:30:18+00:00",
    "next_ping": null,
    "history": [
        {
            "timestamp": "2020-03-23T23:30:18.767Z",
            "status": 1
        },
        {
            "timestamp": "2020-03-25T00:30:18.767Z",
            "status": 0
        }
    ],
    "unique_key": "e855898bebff1756cde7c571319d877d07a38dab",
    "timeout": 86400
}
```

I know that this will require `sendalerts` to be running, but I expect this is enabled for all installations (else the project won't be particularly useful!!).

@cuu508, can you see any other shortcomings of my proposed solution?

If not, I'll write up some documentation and put together a PR.

**Comment by cuu508** [comment]
Great, I was going to suggest to look at the `Flip` model!

A few suggestions on naming.

The `status` field has string values, the `history.status` fields have numeric values. This breaks consistency, perhaps we could use `"up": 0` and `"up": 1` fields (similar to the Prometheus endpoint's output).

The `?history=` parameter isn't very descriptive ("history of what?"). Perhaps rename it to `?flip_log=`. This would leave space for including e.g. `?ping_log=` in the future.

The `?history` parameter accepts values like 6h, 2d, 1w. They look good, but perhaps in the initial implementation we can use simple number-of-seconds instead:
* a plain number of seconds would be easier to validate. I think the current implementation would crash with `?history=abcd` for example. 
* consistent with other API calls that use number of seconds for `timeout` and `grace`

Putting together, how about:

```
GET /api/v1/checks/?flip_log=3600

{
    "name": "My First Check (copy)",
    "tags": "",
    "desc": "",
    "grace": 3600,
    "n_pings": 1,
    "status": "down",
    "last_ping": "2020-03-23T23:30:18+00:00",
    "next_ping": null,
    "flips": [
        {
            "timestamp": "2020-03-23T23:30:18.767Z",
            "up": 1
        },
        {
            "timestamp": "2020-03-25T00:30:18.767Z",
            "up": 0
        }
    ],
    "unique_key": "e855898bebff1756cde7c571319d877d07a38dab",
    "timeout": 86400
}
```

**Comment by alex-phillips** [comment]
This would be great. Is this something to be considered merging into the main project?

**Comment by jameskirsop** [comment]
Yes - but I've not had the time in the past couple of weeks to attend to
completing the PR. I'm hoping to get to it soon.

**Comment by jameskirsop** [comment]
In light of the new `pings` endpoint, I've built a new `flips` endpoint (#371) that returns the history of flips in a similar format, with a couple of query string parameters to determine the time range that you want to retrieve history for. This range feature is important for integration with products like Grafana which allow you go back in time to view data for a defined span of time.

Returning results using this new method means that my Grafana data source will need to make two calls to get all the necessary info for a check, but I think it's better on Healthchecks side to stay consistent in light of the new `pings` endpoint.

Changes for review are at the [same branch as before](https://github.com/jameskirsop/healthchecks/tree/return-single-history). I'll write some tests and do a PR in if we're happy with this.

**Comment by cuu508** [comment]
Adding a `/flips/` endpoint sounds good to me!

Would you be able to make a preliminary PR for reviewing?

**Comment by cuu508** [comment]
I edited and merged #381. It is now also deployed to https://healthchecks.io
